libcudacxx/include/cuda/std/__atomic/order.h

```diff
@@ -0,0 +1,147 @@
+
+#include <cuda/std/detail/__config>
+
+#include <cuda/std/__type_traits/underlying_type.h>
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+#define _LIBCUDACXX_CHECK_STORE_MEMORY_ORDER(__m)                                              \
+  _LIBCUDACXX_DIAGNOSE_WARNING(                                                                \
+    __m == memory_order_consume || __m == memory_order_acquire || __m == memory_order_acq_rel, \
+    "memory order argument to atomic operation is invalid")
+
+#define _LIBCUDACXX_CHECK_LOAD_MEMORY_ORDER(__m)                                           \
+  _LIBCUDACXX_DIAGNOSE_WARNING(__m == memory_order_release || __m == memory_order_acq_rel, \
+                               "memory order argument to atomic operation is invalid")
+
+#define _LIBCUDACXX_CHECK_EXCHANGE_MEMORY_ORDER(__m, __f)                                  \
+  _LIBCUDACXX_DIAGNOSE_WARNING(__f == memory_order_release || __f == memory_order_acq_rel, \
+                               "memory order argument to atomic operation is invalid")
+
+#ifndef __ATOMIC_RELAXED
+#  define __ATOMIC_RELAXED 0
+#  define __ATOMIC_CONSUME 1
+#  define __ATOMIC_ACQUIRE 2
+#  define __ATOMIC_RELEASE 3
+#  define __ATOMIC_ACQ_REL 4
+#  define __ATOMIC_SEQ_CST 5
+#endif //__ATOMIC_RELAXED
+
+// Figure out what the underlying type for `memory_order` would be if it were
+// declared as an unscoped enum (accounting for -fshort-enums). Use this result
+// to pin the underlying type in C++20.
+enum __legacy_memory_order
+{
+  __mo_relaxed,
+  __mo_consume,
+  __mo_acquire,
+  __mo_release,
+  __mo_acq_rel,
+  __mo_seq_cst
+};
+
+typedef underlying_type<__legacy_memory_order>::type __memory_order_underlying_t;
+
+#if _CCCL_STD_VER > 2017
+
+enum class memory_order : __memory_order_underlying_t
+{
+  relaxed = __mo_relaxed,
+  consume = __mo_consume,
+  acquire = __mo_acquire,
+  release = __mo_release,
+  acq_rel = __mo_acq_rel,
+  seq_cst = __mo_seq_cst
+};
+
+inline constexpr auto memory_order_relaxed = memory_order::relaxed;
+inline constexpr auto memory_order_consume = memory_order::consume;
+inline constexpr auto memory_order_acquire = memory_order::acquire;
+inline constexpr auto memory_order_release = memory_order::release;
+inline constexpr auto memory_order_acq_rel = memory_order::acq_rel;
+inline constexpr auto memory_order_seq_cst = memory_order::seq_cst;
+
+#else
+
+typedef enum memory_order
+{
+  memory_order_relaxed = __mo_relaxed,
+  memory_order_consume = __mo_consume,
+  memory_order_acquire = __mo_acquire,
+  memory_order_release = __mo_release,
+  memory_order_acq_rel = __mo_acq_rel,
+  memory_order_seq_cst = __mo_seq_cst,
+} memory_order;
+
+#endif // _CCCL_STD_VER > 2017
+
+_CCCL_HOST_DEVICE inline int __stronger_order_cuda(int __a, int __b)
+{
+  int const __max = __a > __b ? __a : __b;
+  if (__max != __ATOMIC_RELEASE)
+  {
+    return __max;
+  }
+  static int const __xform[] = {__ATOMIC_RELEASE, __ATOMIC_ACQ_REL, __ATOMIC_ACQ_REL, __ATOMIC_RELEASE};
```

Comment:
Could we make this a constexpr variable?
```suggestion
  constexpr int __xform[] = {__ATOMIC_RELEASE, __ATOMIC_ACQ_REL, __ATOMIC_ACQ_REL, __ATOMIC_RELEASE};
```